nemo/collections/asr/parts/submodules/multichannel_modules.py

```diff
@@ -0,0 +1,778 @@
+    hard reference.
+
+    Args:
+        hard: If true, use hard estimate of ref channel.
+            If false, use a soft estimate across channels.
+        hard_use_grad: Use straight-through estimator for
+            the gradient.
+        subband_weighting: If true, use subband weighting when
+            adding across subband SNRs. If false, use average
+            across subbands.
+
+    References:
+        Boeddeker et al, Front-End Processing for the CHiME-5 Dinner Party Scenario, 2018
+    """
+
+    def __init__(
+        self,
+        hard: bool = True,
+        hard_use_grad: bool = True,
+        subband_weighting: bool = False,
+        num_subbands: Optional[int] = None,
+        eps: float = 1e-8,
+    ):
+        super().__init__()
+
+        self.hard = hard
+        self.hard_use_grad = hard_use_grad
+        self.subband_weighting = subband_weighting
+        self.eps = eps
+
+        if subband_weighting and num_subbands is None:
+            raise ValueError(f'Number of subbands must be provided when using subband_weighting={subband_weighting}.')
+        # Subband weighting
+        self.weight_s = torch.nn.Parameter(torch.ones(num_subbands)) if subband_weighting else None
+        self.weight_n = torch.nn.Parameter(torch.ones(num_subbands)) if subband_weighting else None
+
+        logging.debug('Initialized %s', self.__class__.__name__)
+        logging.debug('\thard:              %d', self.hard)
+        logging.debug('\thard_use_grad:     %d', self.hard_use_grad)
+        logging.debug('\tsubband_weighting: %d', self.subband_weighting)
+        logging.debug('\tnum_subbands:      %s', num_subbands)
+        logging.debug('\teps:               %e', self.eps)
+
+    @property
+    def input_types(self):
+        """Returns definitions of module input types
+        """
+        return {
+            'W': NeuralType(('B', 'D', 'C', 'C'), SpectrogramType()),
+            'psd_s': NeuralType(('B', 'D', 'C', 'C'), SpectrogramType()),
+            'psd_n': NeuralType(('B', 'D', 'C', 'C'), SpectrogramType()),
+        }
+
+    @property
+    def output_types(self):
+        """Returns definitions of module output types
+        """
+        return {
+            'output': NeuralType(('B', 'C'), FloatType()),
+        }
+
+    @typecheck()
+    def forward(self, W: torch.Tensor, psd_s: torch.Tensor, psd_n: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            W: Multichannel input multichannel output filter, shape (B, F, C, M), where
+               C is the number of input channels and M is the number of output channels
+            psd_s: Covariance for the signal, shape (B, F, C, C)
+            psd_n: Covariance for the noise, shape (B, F, C, C)
+
+        Returns:
+            One-hot or soft reference channel, shape (B, M)
+        """
+        if self.subband_weighting:
+            # (B, F, M)
+            pow_s = torch.einsum('...jm,...jk,...km->...m', W.conj(), psd_s, W).abs()
```

Comment:
Just a note if you ever intend to export to onnx in the future, it doesn't have great support of einsum